x-pack/plugins/rule_registry/server/utils/create_persistence_rule_type_wrapper.ts

```diff
@@ -146,6 +181,172 @@ export const createPersistenceRuleTypeWrapper: CreatePersistenceRuleTypeWrapper
+                // so future versions of Kibana may read 8.7.0 version alerts and need to update them
+                const response = await ruleDataClient
+                  .getReader({ namespace: options.spaceId })
+                  .search<typeof suppressionAlertSearchRequest, AlertWithSuppressionFields870<{}>>(
+                    suppressionAlertSearchRequest
+                  );
+
+                const existingAlertsByInstanceId = response.hits.hits.reduce<
+                  Record<string, estypes.SearchHit<AlertWithSuppressionFields870<{}>>>
+                >((acc, hit) => {
+                  acc[hit._source['kibana.alert.instance.id']] = hit;
+                  return acc;
+                }, {});
+
+                const [duplicateAlerts, newAlerts] = partition(
+                  alerts,
+                  (alert) =>
+                    existingAlertsByInstanceId[alert._source['kibana.alert.instance.id']] != null
+                );
+
+                const duplicateAlertUpdates = duplicateAlerts.flatMap((alert) => {
+                  const existingAlert =
+                    existingAlertsByInstanceId[alert._source['kibana.alert.instance.id']];
+                  const existingDocsCount =
+                    existingAlert._source?.[ALERT_SUPPRESSION_DOCS_COUNT] ?? 0;
+                  return [
+                    {
+                      update: {
+                        _id: existingAlert._id,
+                        _index: existingAlert._index,
+                        require_alias: false,
+                      },
+                    },
+                    {
+                      doc: {
+                        [ALERT_LAST_DETECTED]: currentTimeOverride ?? new Date(),
+                        [ALERT_SUPPRESSION_END]: alert._source[ALERT_SUPPRESSION_END],
+                        [ALERT_SUPPRESSION_DOCS_COUNT]:
+                          existingDocsCount + alert._source[ALERT_SUPPRESSION_DOCS_COUNT] + 1,
+                      },
+                    },
+                  ];
+                });
+
+                let enrichedAlerts = newAlerts;
+
+                if (enrichAlerts) {
+                  try {
+                    enrichedAlerts = await enrichAlerts(enrichedAlerts, {
+                      spaceId: options.spaceId,
+                    });
+                  } catch (e) {
+                    logger.debug('Enrichments failed');
+                  }
+                }
+
+                const augmentedAlerts = augmentAlerts({
+                  alerts: enrichedAlerts,
+                  options,
+                  kibanaVersion: ruleDataClient.kibanaVersion,
+                  includeLastDetected: true,
+                  currentTimeOverride,
+                });
+
+                const bulkResponse = await ruleDataClientWriter.bulk({
+                  body: [...duplicateAlertUpdates, ...mapAlertsToBulkCreate(augmentedAlerts)],
+                  refresh: true,
```

Comment:
Should this be `wait_for` instead of `true`?

```suggestion
                  refresh: 'wait_for`,
```

Reply:
There are advantages and disadvantages for both options. IMO we should guarantee that the alerts index is refreshed before the rule execution is complete, which eliminates `refresh: false` as an option up front (this is most important for rule preview, but also guarantees that real rule executions will update alerts correctly). The downside with `wait_for` is it can add a significant amount of time to the rule execution which would negatively affect rule throughput (up to 1 additional second, with the default setting [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#index-refresh-interval-setting). `true` shifts the burden to ES but could cause a lot of refreshes with many simultaneous rule executions.

Perhaps we can decrease the refresh interval on the alerts index and use `wait_for` here if it's likely that we'll have many rules executing and explicitly refreshing the alerts index simultaneously?

Reply:
I chose `refresh: true` initially with the thinking that the index already refreshes every second by default, so adding explicit refreshes per suppression rule execution is unlikely to significantly increase the overall rate of refreshes ES has to do on the alert index.

Reply:
Not a problem, I agree that we need the data refreshed before the rule execution is complete. Soon the framework will rely on the data being refreshed to apply conditional actions, maintenance windows, etc..

We may have to change this to `wait_for` down the line depending on the details of stateless Elasticsearch. We can keep it to true for now and change down the line as needed 👍